File: backend/app/models/schema.py
"""Pydantic models for Schema Generator feature."""

from datetime import datetime
from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field


class FieldType(str, Enum):
    """Supported field types in extracted schemas."""
    TEXT = "text"
    NUMBER = "number"
    DATE = "date"
    CHOICE = "choice"
    GRADED_CHOICE = "graded_choice"
    BOOLEAN = "boolean"


class GradeOption(BaseModel):
    """Grade option for graded choice fields."""
    value: int = Field(..., description="Numeric grade value (e.g., 5, 7, 9)")
    label: str = Field(..., description="Grade label description")
    label_id: Optional[str] = Field(None, description="Indonesian label (bilingual support)")


class SchemaField(BaseModel):
    """Individual field in extracted schema."""
    id: str = Field(..., description="Unique field identifier")
    label: str = Field(..., description="Field label (English)")
    label_id: Optional[str] = Field(None, description="Field label (Indonesian)")
    field_type: FieldType = Field(..., description="Field data type")
    required: bool = Field(True, description="Whether field is required")
    default_value: Optional[str] = Field(None, description="Default value if any")
    validation_rules: Optional[dict] = Field(None, description="Validation rules (min, max, pattern)")
    options: Optional[list[GradeOption]] = Field(None, description="Options for choice/graded_choice fields")
    unit: Optional[str] = Field(None, description="Unit of measurement (e.g., 'kg', 'C')")


class SchemaCriterion(BaseModel):
    """Criterion within a section (e.g., 'Appearance' with grades 5/7/9)."""
    id: str = Field(..., description="Unique criterion identifier")
    label: str = Field(..., description="Criterion label (English)")
    label_id: Optional[str] = Field(None, description="Criterion label (Indonesian)")
    grades: list[GradeOption] = Field(default_factory=list, description="Available grades for this criterion")


class SchemaSection(BaseModel):
    """Section containing multiple criteria (e.g., 'Frozen', 'Thawing')."""
    id: str = Field(..., description="Unique section identifier")
    label: str = Field(..., description="Section label (English)")
    label_id: Optional[str] = Field(None, description="Section label (Indonesian)")
    criteria: list[SchemaCriterion] = Field(default_factory=list, description="Criteria in this section")


class ExtractedSchemaStructure(BaseModel):
    """Complete extracted schema structure representing one sample's data."""
    per_sample_fields: list[SchemaField] = Field(
        default_factory=list,
        description="Fields that apply to each sample (temperature, weight)"
    )
    sections: list[SchemaSection] = Field(
        default_factory=list,
        description="Sections containing criteria with grades"
    )
    batch_metadata_fields: list[SchemaField] = Field(
        default_factory=list,
        description="Batch-level fields (supplier, date, lot number)"
    )
    validation_rules: Optional[dict] = Field(
        None,
        description="Global validation rules from form notes/footers"
    )


class ExtractionMetadata(BaseModel):
    """Metadata about the extraction process."""
    source_file: str = Field(..., description="Original filename")
    source_file_size: int = Field(..., description="File size in bytes")
    model_used: str = Field(..., description="LLM model used for extraction")
    confidence_score: float = Field(..., ge=0.0, le=1.0, description="Extraction confidence (0-1)")
    extraction_timestamp: datetime = Field(default_factory=datetime.utcnow)
    processing_time_ms: Optional[int] = Field(None, description="Processing time in milliseconds")


# Request/Response Models

class SchemaExtractionRequest(BaseModel):
    """Request for schema extraction (file is sent as multipart/form-data)."""
    schema_name: str = Field(..., min_length=1, max_length=255, description="Name for the extracted schema")


class SchemaExtractionResponse(BaseModel):
    """Response from schema extraction endpoint."""
    extracted_schema: ExtractedSchemaStructure = Field(..., description="Extracted schema structure")
    confidence_score: float = Field(..., ge=0.0, le=1.0, description="Overall extraction confidence")
    extraction_metadata: ExtractionMetadata = Field(..., description="Extraction process metadata")
    warnings: list[str] = Field(default_factory=list, description="Extraction warnings or notes")


class SchemaCreateRequest(BaseModel):
    """Request to save a schema."""
    form_code: str = Field(..., min_length=1, max_length=50, description="Form code identifier")
    form_name: str = Field(..., min_length=1, max_length=255, description="Human-readable form name")
    category: Optional[str] = Field(None, max_length=100, description="Form category")
    schema_definition: ExtractedSchemaStructure = Field(..., description="Schema structure")
    extraction_metadata: Optional[ExtractionMetadata] = Field(None, description="Extraction metadata if AI-generated")


class SchemaUpdateRequest(BaseModel):
    """Request to update an existing schema (creates new version)."""
    schema_definition: ExtractedSchemaStructure = Field(..., description="Updated schema structure")
    update_reason: Optional[str] = Field(None, description="Reason for update")


class SchemaResponse(BaseModel):
    """Schema response with full details."""
    id: str
    form_code: str
    form_name: str
    category: Optional[str]
    version: str
    version_number: int
    schema_definition: ExtractedSchemaStructure
    status: str
    extraction_metadata: Optional[ExtractionMetadata]
    created_at: datetime
    updated_at: datetime


class SchemaListItem(BaseModel):
    """Lightweight schema item for list views."""
    id: str
    form_code: str
    form_name: str
    category: Optional[str]
    version: str
    version_number: int
    status: str
    created_at: datetime


class SchemaListResponse(BaseModel):
    """Paginated list of schemas."""
    schemas: list[SchemaListItem]
    total: int
    page: int
    page_size: int

File: web/src/types/schema.ts
/**
 * TypeScript types for Schema Generator feature.
 * Mirrors backend/app/models/schema.py
 */

export type FieldType = 'text' | 'number' | 'date' | 'choice' | 'graded_choice' | 'boolean';

export interface GradeOption {
  value: number;
  label: string;
  label_id?: string;
}

export interface SchemaField {
  id: string;
  label: string;
  label_id?: string;
  field_type: FieldType;
  required: boolean;
  default_value?: string;
  validation_rules?: Record<string, unknown>;
  options?: GradeOption[];
  unit?: string;
}

export interface SchemaCriterion {
  id: string;
  label: string;
  label_id?: string;
  grades: GradeOption[];
}

export interface SchemaSection {
  id: string;
  label: string;
  label_id?: string;
  criteria: SchemaCriterion[];
}

export interface ExtractedSchemaStructure {
  per_sample_fields: SchemaField[];
  sections: SchemaSection[];
  batch_metadata_fields: SchemaField[];
  validation_rules?: Record<string, unknown>;
}

export interface ExtractionMetadata {
  source_file: string;
  source_file_size: number;
  model_used: string;
  confidence_score: number;
  extraction_timestamp: string;
  processing_time_ms?: number;
}

export interface FilePreviewInfo {
  url: string;
  filename: string;
  size: number;
  mime_type: string;
  page_count?: number;
}

// API Request/Response Types

export interface SchemaExtractionRequest {
  schema_name: string;
}

export interface SchemaExtractionResponse {
  extracted_schema: ExtractedSchemaStructure;
  confidence_score: number;
  extraction_metadata: ExtractionMetadata;
  warnings: string[];
  file_preview?: FilePreviewInfo | null;
}

export interface SchemaCreateRequest {
  form_code: string;
  form_name: string;
  category?: string;
  schema_definition: ExtractedSchemaStructure;
  extraction_metadata?: ExtractionMetadata;
}

export interface SchemaUpdateRequest {
  schema_definition: ExtractedSchemaStructure;
  update_reason?: string;
}

export interface SchemaResponse {
  id: string;
  form_code: string;
  form_name: string;
  category?: string;
  version: string;
  version_number: number;
  schema_definition: ExtractedSchemaStructure;
  status: string;
  extraction_metadata?: ExtractionMetadata;
  created_at: string;
  updated_at: string;
}

export interface SchemaListItem {
  id: string;
  form_code: string;
  form_name: string;
  category?: string;
  version: string;
  version_number: number;
  status: string;
  created_at: string;
  has_source_document: boolean;
}

export interface SchemaListResponse {
  schemas: SchemaListItem[];
  total: number;
  page: number;
  page_size: number;
}

// Source Document Types

export interface SourceDocumentInfo {
  path: string;
  filename: string;
  size: number;
  mime_type: string;
  url: string;
}

// Bulk Operations Types

export interface BulkArchiveRequest {
  schema_ids: string[];
}

export interface BulkArchiveResponse {
  archived_count: number;
  failed_ids: string[];
  errors: string[];
}
